azavea/raster-foundry
# WIP: Add diagram node preview and comparison

## Overview

This PR does many things:
  - Refactors context menu code to render correctly in spite of angular's attempts to do the opposite
  - Adds the side-by-side leaflet control (see notes)
  - Adds a mousetip (a tooltip but travels with mouse) service
  - Adds a preview map that will show a single node's output or compare two nodes' output

### Demo

<img width="1280" alt="screen shot 2017-01-18 at 12 31 02 pm" src="https://cloud.githubusercontent.com/assets/2442245/22075477/2170379c-dd7a-11e6-9983-3ec0c0faec42.png">
<img width="1278" alt="screen shot 2017-01-18 at 12 31 11 pm" src="https://cloud.githubusercontent.com/assets/2442245/22075476/216f2adc-dd7a-11e6-9985-8038cd7f2e0b.png">
<img width="1278" alt="screen shot 2017-01-18 at 12 31 23 pm" src="https://cloud.githubusercontent.com/assets/2442245/22075478/21725374-dd7a-11e6-92e2-320a0f7202ab.png">
<img width="1276" alt="screen shot 2017-01-18 at 12 31 36 pm" src="https://cloud.githubusercontent.com/assets/2442245/22075480/2173d4e2-dd7a-11e6-8959-cd301e1434d9.png">
<img width="1278" alt="screen shot 2017-01-18 at 12 31 44 pm" src="https://cloud.githubusercontent.com/assets/2442245/22075479/217371be-dd7a-11e6-875f-d263661e6062.png">


### Notes

The leaflet side-by-side control is developed to be used with browserify and attempting to use that within the webpack context blew up leaflet. The only way I managed to get the control to cooperate was to use a special build of the side-by-side control and put it directly into the codebase. It's ugly, but it works.

@designmatty, the preview slider is not styled, but can be if we strip out the stylesheet injected by the `leaflet-side-by-side.js` file. Also, the previewed nodes' names are displayed at the bottom of the map, but the style does not reflect what was in the mocks. The map attribution now has the same style as the mocked-up node names, so I thought we could change it for now and address this issue in a future style clean up.

When in comparison mode, it would probably be best to disable dragging of nodes and have the cursor reflect selection (`cursor: pointer`) when hovering a node.

After clicking the 'Compare to...' option, if the user clicks 'Cancel' or clicks the diagram background, we cancel the comparison process.

Currently, the PR is WIP as we need endpoints to truly compare or preview individual nodes.

## Testing Instructions

 * Go to the lab
 * Select a node, see that the context menu shows up as expected
 * Click 'View Output', see that the map comes up in single display mode
 * Exit the preview, and click 'Compare to...' within the context menu and select a node to compare to, see that the map comes up in comparison mode.
